src/goto-programs/goto_program.cpp

```diff
@@ -235,23 +237,23 @@ void parse_lhs_read(const exprt &src, std::list<exprt> &dest)
 {
   if(src.id()==ID_dereference)
   {
-    assert(src.operands().size()==1);
+    PRECONDITION(src.operands().size() == 1);
```

Comment:
I figured this would go very well with the changes suggested in #2902, so I went ahead to change all of these, by removing the duplicate preconditions and converting the untyped expressions into the correct types with the typecast expressions.